src/google/protobuf/compiler/java/kotlin_main.cc

```diff
@@ -0,0 +1,37 @@
+// Protocol Buffers - Google's data interchange format
+// Copyright 2008 Google Inc.  All rights reserved.
+// https://developers.google.com/protocol-buffers/
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//     * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "google/protobuf/compiler/java/kotlin_generator.h"
+#include "google/protobuf/compiler/plugin.h"
+
+int main(int argc, char *argv[]) {
+  google::protobuf::compiler::java::KotlinGenerator generator;
```

Comment:
doesn't this also need the JavaGenerator to be useful?

Reply:
Sorry, could you please elaborate if possible? Right now I have a separate entrypoint for Java in `src/google/protobuf/compiler/java/main.cc`, and this entrypoint here using only KotlinGenerator matches what we already do to make an unofficial `protoc-gen-kotlin`.

Reply:
I have two distinct directions of concern/thought here.

1) I am imagining a hypothetical user that has installed `protoc-gen-kotlin`.  They are able to generate kotlin code, but that code depends on a java library that they are unable to generate.  Seems like not a great situation.

2) A user may have different versions of `protoc-gen-kotlin` and `protoc-gen-java`.  I literally don't know if their generated kotlin stubs will work.  I am not certain how tight the coupling is between these two and don't relish finding out the hard way.

Reply:
> A user may have different versions of protoc-gen-kotlin and protoc-gen-java. I literally don't know if their generated kotlin stubs will work. I am not certain how tight the coupling is between these two and don't relish finding out the hard way.

I don't think this is any different than the rest of the Protobuf ecosystem though - you could have different versions of dependent plugins, say a Protobuf and grpc plugin, and it may or may not cause an issue. I can see the integration here may be tighter, but I'd think that it's an orthogonal issue. I'd love to have a way to declare library and plugin dependencies for plugins, but it's not part of the plugin API at the moment

Reply:
(1) is genuinely different, we don't have a way to end up in this state right now.

(2) has existing issues, but they require having two distinct full copies of `protoc` installed, which is not nearly as easy as having two different binaries installed.

Reply:
Perhaps we're talking about two different things here - my argument is that neither (1) or (2) is any different from the rest of the plugin ecosystem, and I don't know if there's a reason to special case kotlin and Java. As an example, if you generate code with protoc-gen-go-grpc, you need generated code from protoc-gen-go, but if you don't have protoc-gen-go installed, you can't generate it.

If there's a kotlin/Java-specific solution to this though, that's great. But I think my argument is: massively in favor of dependencies between plugins (we actually represent the dependencies in BSR), but not sure builtins are any different than the rest of the ecosystem, and not sure if this is the place to have special-case guardrails. But I may be missing something massive here (wouldn't be the first time), apologies if so.